dj_gui_api_server/DJConnector.py

```diff
@@ -109,9 +111,46 @@ def fetch_tuples(jwt_payload: dict, schema_name: str, table_name: str):
         :type schema_name: str
         :param table_name: Table name under the given schema; must be in camel case
         :type table_name: str
-        :return: List of tuples in dict form
-        :rtype: list
+        :param restriction: Sequence of filter cards with attribute_name, operation, value
+            defined, defaults to []
+        :type restriction: list, optional
+        :param limit: Max number of records to return, defaults to 1000
+        :type limit: int, optional
+        :param page: Page number to return, defaults to 1
+        :type page: int, optional
+        :param order: Sequence to order records, defaults to ['KEY ASC'].
+            See :class:`datajoint.fetch.Fetch` for more info.
+        :type order: list, optional
+        :return: Records in dict form and the total number of records that can be paged
+        :rtype: tuple
         """
+        def resolve_expression(all_restrictions: list,
+                               query: QueryExpression) -> QueryExpression:
+            if not all_restrictions:
+                return query & dict()
+            else:
+                current_restriction = all_restrictions[0]
+                if current_restriction['operation'] in ('>', '<', '>=', '<='):
+                    operation = current_restriction['operation']
+                elif current_restriction['value'] is None:
+                    operation = (' IS ' if current_restriction['operation'] == '='
+                                 else ' IS NOT ')
+                else:
+                    operation = current_restriction['operation']
+
+                if (isinstance(current_restriction['value'], str) and
+                        not current_restriction['value'].isnumeric()):
+                    value = f"'{current_restriction['value']}'"
+                else:
+                    value = ('NULL' if current_restriction['value'] is None
+                             else current_restriction['value'])
+
+                current_restriction = f"""{current_restriction[
+                    'attributeName']}{operation}{value}"""
+                return (query & current_restriction if len(all_restrictions) == 1
+                        else resolve_expression(all_restrictions[1:],
+                                                query & current_restriction))
+
```

Comment:
Not really a fan of function inside a function... Also I could have recall there is are some performance downside to this apporach of nested functions

Reply:
Recursion in functions are part of standard good practice in Python provided they:

 1. Simplify the logic compared to iterative means
 2. Safer to do
 3. Doesn't grow too large.

It is true that there is some performance drawback compared to the iterative equivalent but it is not as much as you might think (recursion is used quite a bit in Python community b/c of this). The specific challenge here though is that we need to dynamically build an expression from a list of terms. Not only is the logic more complex but to do this in an iterative means would require us creating a string with the complete Python expression and run it by way of `exec`. If a user is clever enough, this could be used to compromise the system (similar to SQL injection but at the Python backend level!).

That said though, you motivated me into rethinking this which I stumbled upon `reduce` which further optimized and simplified things. :smiley: Recursion has been removed since it is now being handled within `reduce` built-in. This new update should address your concerns.

Reply:
Hmm, that is a new feature. Cool find on reduce.